python/cuml/metrics/pairwise_kernels.py

```diff
@@ -0,0 +1,278 @@
+    array.
+    If Y is given (default is None), then the returned matrix is the pairwise
+    kernel between the arrays from both X and Y.
+    Valid values for metric are:
+        ['additive_chi2', 'chi2', 'linear', 'poly', 'polynomial', 'rbf',
+        'laplacian', 'sigmoid', 'cosine']
+    Parameters
+    ----------
+    X : Dense matrix (device or host) of shape (n_samples_X, n_samples_X) or \
+            (n_samples_X, n_features)
+        Array of pairwise kernels between samples, or a feature array.
+        The shape of the array should be (n_samples_X, n_samples_X) if
+        metric == "precomputed" and (n_samples_X, n_features) otherwise.
+        Acceptable formats: cuDF DataFrame, NumPy ndarray, Numba device
+        ndarray, cuda array interface compliant array like CuPy
+    Y : Dense matrix (device or host) of shape (n_samples_Y, n_features),
+        default=None
+        A second feature array only if X has shape (n_samples_X, n_features).
+        Acceptable formats: cuDF DataFrame, NumPy ndarray, Numba device
+        ndarray, cuda array interface compliant array like CuPy
+    metric : str or callable (numba device function), default="linear"
+        The metric to use when calculating kernel between instances in a
+        feature array.
+        If metric is "precomputed", X is assumed to be a kernel matrix.
+        Alternatively, if metric is a callable function, it is called on each
+        pair of instances (rows) and the resulting value recorded. The callable
+        should take two rows from X as input and return the corresponding
+        kernel value as a single number.
+    filter_params : bool, default=False
+        Whether to filter invalid parameters or not.
+    convert_dtype : bool, optional (default = True)
+        When set to True, the method will, when necessary, convert
+        Y to be the same data type as X if they differ. This
+        will increase memory used for the method.
+    **kwds : optional keyword parameters
+        Any further parameters are passed directly to the kernel function.
+    Returns
+    -------
+    K : ndarray of shape (n_samples_X, n_samples_X) or \
+            (n_samples_X, n_samples_Y)
+        A kernel matrix K such that K_{i, j} is the kernel between the
+        ith and jth vectors of the given matrix X, if Y is None.
+        If Y is not None, then K_{i, j} is the kernel between the ith array
+        from X and the jth array from Y.
+    Notes
+    -----
+    If metric is 'precomputed', Y is ignored and X is returned.
+
+    Examples
+    --------
+
+    .. code-block:: python
+
+        import cupy as cp
+        from cuml.metrics import pairwise_kernels
+        from numba import cuda
+        import math
+
+        X = cp.array([[2, 3], [3, 5], [5, 8]])
+        Y = cp.array([[1, 0], [2, 1]])
+
+        pairwise_kernels(X, Y, metric='linear')
+
+        @cuda.jit(device=True)
+        def custom_rbf_kernel(x, y, gamma=None):
```

Comment:
I very much like the ability to quickly build custom kernels. Have you done any profiling / benchmarking of this against the `cuml.metrics.pairwise_distances` API? I'm mostly curious to know the gap between the two, and whether there's a perf hit for the different memory access patterns.

Reply:
It would be interesting to see the difference, but for now it doesn't really matter as the matrix inversion dominates computation time. It could be 5 times slower than the cuda version and we won't see any real difference in end to end time.

The bigger disadvantage of this approach for me has been jit compile time.  It's in the range of a few hundred ms, which I think is reasonable.

Reply:
No worries. My question isn't about this algorithm in particular. It's been on our todo list for quite awhile to see how performant it would be to allow users to implement custom pairwise distance measures in Numba.